### ISSUE TITLE
repeatedly detection of G-clef at an eights note with flag down

### ISSUE THREAD

**Issue by Bacchushlg** [Audiveris/audiveris#83]
![beatles - if i fell in love 2](https://user-images.githubusercontent.com/33522630/36413142-683d8d64-161d-11e8-8d71-9dbbdde275e9.jpg)

In this score I get twice a G-clef detected at an eighth note. For me it's not obvious, why.
At another position Audiveris detects a half note instead.

If I remove the wrong clefs and half notes and export, then import to MuseScore, the clefs appear in the score?!?

**Comment by hbitteur** [comment]
![image](https://user-images.githubusercontent.com/22053634/36437183-97b738da-1666-11e8-80d7-2d4a7f36704a.png)
Here is the "explanation", based on snapshot taken at end of STEMS step:
- The stem on left and the top black head were correctly detected.
- However, a (false) stem was also detected on the right side, together with a (false) void head near the bottom on this false stem.

And a head can be connected to 2 stems only in a very canonical situation (stem on bottom-left and stem on top-right). All the other cases are illegal, so some items have to be removed.
The top note head appears wrongly connected to left and right stems, and it got pruned off from the right stem. To see this in action, set this head as VIP (from InterBoard), and you'll see what is happening to the VIP entity when running REDUCTION step.

Here, the false stem should have been discarded instead of the correct black head, which was found on the wrong side of the false stem. I need further investigations.

**Comment by hbitteur** [comment]
![image](https://user-images.githubusercontent.com/22053634/36442813-3a53c9ae-1677-11e8-9e08-4e320bbfa0e4.png)
Bad news:
At some point in REDUCTION step, we have left stem with 2 heads (top & bottom) and right stem with 1 
head (bottom). Note the bottom head (highleghted) is linked to both left & right stems, which is not a legal case, but OMR engine has not been able to cut this.
What is happening now is that the left stem, because of the link with the bottom void head, exhibits a too short tail and is thus discarded. And only the false right stem with the false void head survive...

**Comment by Bacchushlg** [comment]
maybe it would be a good idea to split the step reduction into 2 steps.
First step works like now, but keeps some alternatives "in mind"
Second steps then runs after step symbols and cleans up the rest.

Actually the flags are detected in step symbol rather late. So in our case, due to reduction, the flag has no chance to be detected, because the stem is already removed.

**Comment by Bacchushlg** [comment]
what about the last comment in the original report: the presence of the manually removed clef in the export. Shall I open a separate issue for this?

**Comment by hbitteur** [comment]
Regarding the clef, a removed item should not persist in the project or export file, it's a bug, yes please fill a report for it.

Regarding reductions, one is performed in REDUCTION step, another one in LINKS step. The former addresses hote heads, stems and beams, the latter all the other symbols, including flags.
This split took a long time and many design modifications back and forth, it works rather well today, I don't want to change it again. Remember, this 5.1 release was meant to be a small release with a basic UI handling enough corrections to allow the training of the new classifiers coming. 

Instead, I'm thinking of improving the reduction algorithm. Back to our example, what triggered the removal of the (correct) stem on the left was the fact that the remaining length of the stem tail was too short. But this small tail length resulted from the presence of the (incorrect) void head on the (correct) stem. The problem is that this void head has a good contextual grade (>= 0.5). Fortunately, it is  still lower than the contextual grade of the (correct) black head. The black head needs the stem, so rather than removing the stem, the algorithm should have:
1. Cut the link between this stem and the (incorrect) void head,
2. Removed the void head as well because it overlaps the stem needed by the black head whose grade is higher than void head grade.

In other words, there should be  an exclusion between the black head and the void head, rather difficult to notice because it goes through the intermediate needed stem.
And in case you wonder, there are some cases where we can legally have both a void head and a black head on a common stem (case where the void head is actually shared with another chord in the opposite direction).

OK, this is a bit complex, I'll think it over. I'm leaving for a couple of days, with nearly no Internet connection. Back beginning of next week, perhaps with a fix.

**Comment by hbitteur** [comment]
Now fixed by commit f6313ac34bd35efdb4a582719bce48c7a99576df which basically inserts exclusion  between a stem and a head if they are not compatible while they overlap.
